File: java/intelijidea/chegg-i19/src/LinkedList.java
import java.util.Iterator;
import java.util.NoSuchElementException;

public class LinkedList<E> implements Iterable<E>{
    private Node head;
    private Node tail;
    private int size;

    LinkedList(){
        this.head = null;
        this.tail = null;
        this.size = 0;
    }

    /**
     *
     * @param data data value to be insert in linked list
     */
    void add(E data){
        Node newNode = new Node(data);
        if(this.head == null){
            this.head = newNode;
            this.tail = newNode;
            this.size++;
            return;
        }
        this.tail.next = newNode;
        this.tail = newNode;
        this.size++;
    }

    /**
     *
     * @param n remove a node from list
     */
    void remove(Node n){
        if(n == this.head){
            this.head = this.head.next;
            this.size--;
            return;
        }
        Node current = this.head.next;
        Node prev = this.head;
        while (current != null){
            if(current == n){
                prev.next = current.next;
                return;
            }
            prev = current;
            current = current.next;
        }
        this.size--;
    }

    @Override
    public Iterator<E> iterator() {
        return new LinkedIterator<>(this.head);
    }


    class Node<E>{
        Node next;
        E data;
        Node(E data, Node next){
            this.data = data;
            this.next = next;
        }
        Node(E data){
            this.data = data;
            this.next = null;
        }
    }

    class LinkedIterator<E> implements Iterator<E>{
        private Node currentNode;

        LinkedIterator(Node<E> head){
            this.currentNode = head;
        }

        @Override
        public boolean hasNext() {
            return this.currentNode != null;
        }

        @Override
        public void remove() {
            if(this.hasNext()){
                LinkedList.this.remove(this.currentNode);
                LinkedList.this.size--;
            }else{
                throw new IllegalStateException();
            }
        }

        @Override
        public E next() {
            if(!this.hasNext()){
                throw new NoSuchElementException();
            }
            E data = (E) this.currentNode.data;
            this.currentNode = this.currentNode.next;
            return data;
        }
    }
}

File: c++/1 Eclipse cpp/chegg-06/src/LinkedList.cpp
/*
 * LinkedList.cpp
 *
 *  Created on: 22-Jun-2020
 *      Author: naren
 */

#include "LinkedList.h"

LinkedList::LinkedList() {
	this->head = NULL;
	this->tail = NULL;
}

void LinkedList::add(int data){
	// if list is empty create new node and initialize head and tail to newNode
	if(isEmpty()){
		Node *newNode = new Node(data);
		this->head = newNode;
		this->tail = newNode;
		return;
	}

	// if data is equal to head data return do nothing
	if(this->head->data == data){
		return;
	}

	Node *curr = this->head->next;
	Node *pre = curr;
	while(curr != NULL){
		if(curr->data == data){
			// swap current node data to previous node data
			curr->data = pre->data;
			pre->data = data;
			return;
		}
		pre = curr;
		curr = curr->next;
	}

	// if node is not exist then add newNode as last node of current linked list
	Node *newNode = new Node(data);
	this->tail->next = newNode;
	this->tail = newNode;
}

void LinkedList::add(Node *newNode){
	this->add(newNode->data);
}


void LinkedList::print(){
	if(this->isEmpty()){
		cout<<"Linked list is empty: "<<endl; // show message if list is empty
		return;
	}

	Node *curr = this->head;
	while(curr != NULL){
		cout<<curr->data<<"->"; // print data of node
		curr = curr->next;
	}

	cout<<"NULL"<<endl;
}

bool LinkedList::isEmpty(){
	if(this->head == NULL){
		return true; // return true if list is empty
	}
	return false; // return false if list is not empty
}

LinkedList::~LinkedList() {
	// free all occupy memory
	delete this->head;
	delete this->tail;
}

